tools/nl/embeddings/finetune.py

```diff
@@ -0,0 +1,258 @@
+
+BATCH_SIZE = 256
+NUM_WARMUP_STEPS = 10
+NUM_EPOCHS = 10
+
+VERY_HIGH_MATCH_SCORE = 0.95
+HIGH_MATCH_SCORE = 0.90
+MEDIUM_HIGH_MATCH_SCORE = 0.85
+
+
+def _upload_to_gcs(ctx: utils.Context,
+                   gcs_path: str,
+                   local_filepath: str,
+                   empty_folder: bool = False) -> None:
+  print(f'Uploading {local_filepath}')
+  blob = ctx.bucket.blob(gcs_path)
+
+  if empty_folder:
+    blob.upload_from_string(
+        '', content_type='application/x-www-form-urlencoded;charset=UTF-8')
+  else:
+    blob.upload_from_filename(local_filepath)
+  print(f'Path in GCS: {gcs_path}')
+
+
+def _make_gcs_model_folder(base_model_name: str) -> str:
+  now = datetime.now()
+
+  month_str = utils.two_digits(now.month)
+  day_str = utils.two_digits(now.day)
+  hour_str = utils.two_digits(now.hour)
+  minute_str = utils.two_digits(now.minute)
+  second_str = utils.two_digits(now.second)
+
+  return f"model_{base_model_name}_{now.year}_{month_str}_{day_str}_{hour_str}_{minute_str}_{second_str}"
+
+
+def _alternatives(autogen_input_filepattern: str,
+                  alternatives_filepattern: str) -> pd.DataFrame:
+
+  df_svs = pd.DataFrame()
+  # Append autogen CSVs if any.
+  autogen_dfs = []
+  for autogen_csv in sorted(glob.glob(autogen_input_filepattern)):
+    print(f'Processing autogen input file: {autogen_csv}')
+    autogen_dfs.append(pd.read_csv(autogen_csv).fillna(""))
+  if autogen_dfs:
+    df_svs = pd.concat([df_svs] + autogen_dfs)
+    df_svs = df_svs.drop_duplicates(subset=utils.DCID_COL)
+
+  # Get alternatives and add to the dataframe.
+  for alt_fp in sorted(glob.glob(alternatives_filepattern)):
+    df_alts = utils.get_local_alternatives(
+        alt_fp, [utils.DCID_COL, utils.ALTERNATIVES_COL])
+    df_svs = utils.merge_dataframes(df_svs, df_alts)
+
+  return df_svs
+
+
+def fine_tune_model(model: Any, df_svs: pd.DataFrame,
+                    df_sentence_pairs: pd.DataFrame):
+  """Fine tuning involves providing pairs of sentences/text with an
+  approximate similar score to a baseline model. These pairs (and scores)
+  are used for further `training` (with CosineSimilarityLoss as the objecive).
+  The end result is that the baseline model's weights get updated based on
+  the `new` training examples (pairs).
+  Effectively, the training examples (pairs and scores) provide additional
+  context to a baseline model about the kinds of associations between
+  sentences/text that we care about. In this case, we use the StatVar name,
+  description and alternatives (human curated and LLM-generated) to create
+  pairs of sentences/text with high similarity scores. We also provide a
+  way to indicate that some pairs of texts/words/sentences should not be
+  associated with each other by assigning very low scores. This allows us to
+  bias the model's weights to accomodate our particular use case better. 
+  """
+
+  training_examples: List[InputExample] = []
```

Comment:
nit: maybe this section could be in a `_generate_training_examples()` helper for readability, so this function can focus on the core finetuning logic :)

Reply:
Added a `_generate_training_examples()` function

Reply:
Also meant the function comments needed to be updated. Done.